Best practices/Testing.md

```diff
@@ -205,12 +207,105 @@ Common patterns for organizing test and fixture files makes it easier to navigat
+
+* If you are trying to find one of several instances of a component, use one of the following strategies (listed here from most to least preferred):
+
+  * If the component makes sense as a separate, named component, extract and find it based on that new component. Use this strategy only when the component encapsulates sufficient logic to make it meaningful as a separate component instead of overloading the parent component with that responsibility.
+  
+    This is frequently the case for modals, which are often better off encapsulating all their logic in a separate component that can be found independently of other modals on the page:
+
+    ```ts
+    // instead of:
+
+    <Modal
+      testID="DeleteModal"
+      title="Delete product?"
+      open={this.state.deleteOpen}
+    >
+      {deleteContents}
+    </Modal>
+    <Modal
+      testID="DuplicateModal"
+      title="Duplicate product?"
+      open={this.state.duplicateOpen}
+    >
+      {duplicateContents}
+    </Modal>
+
+    // make them into separate components:
+
+    <DeleteModal open={this.state.deleteOpen} />
+    <DuplicateModal open={this.state.duplicateOpen} />
+
+    // and, in your test, you can now find based on this component,
+    // shown here using Enzyme:
+
+    expect(myComponent.find(DeleteModal)).toHaveProp('open', true);
+    ```
+  
+  * If the component accepts an `id` property that is meaningful on its own, filter components by that property instead. This is frequently the case for form controls, where omitting `id` properties are allowed but result in non-deterministic IDs being used in their place:
+
+    ```ts
+    // instead of this (using the TextField component from Polaris React):
+
+    <TextField testID="NameField" label="Name" value={this.state.name}>
+    <TextField testID="AgeField" label="Age" value={this.state.age}>
+
+    // use the real ID prop, which is eventually needed to associate
+    // the label field with the input:
+
+    <TextField id="NameField" label="Name" value={this.state.name}>
+    <TextField id="AgeField" label="Age" value={this.state.age}>
+
+    // and, in your test, you can now find based on this property,
+    // shown here using Enzyme:
+
+    expect(myComponent.find(TextField).filter({id: 'NameField'})).toHaveProp('value', '');
+    ```
+
+    To make this approach even easier, our [`enzyme-utilities` package](https://github.com/Shopify/quilt/tree/master/packages/enzyme-utilities) provides a `findById` utility function.
+  
+  * If neither of the above apply, filter the set of matched components based on the properties that are unique to them. If no properties are unique, fall back to the index of the item in the matched set.
```

Comment:
this wording is a bit confusing to me. To confirm, if you can't make it into a separate component, and a unique ID doesn't make sense, then simply find all instances of the component and filter them by some other unique thing like text or a prop. If there are no unique attributes that make sense, then locate the item by its index (ex. the second modal)?

Reply:
Yes, that is exactly correct (text is always a prop of the component, but I wanted to try to not tie this entire section just to React)